include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,782 @@
+  * \brief Contains gap decorator to annotate sequences with gaps using a set.
+  */
+
+#pragma once
+
+#include <limits>
+#include <set>
+#include <tuple>
+#include <type_traits>
+
+#include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/range/container/concept.hpp>
+#include <seqan3/range/detail/random_access_iterator.hpp>
+#include <seqan3/std/view/view_all.hpp>
+#include <seqan3/std/ranges>
+
+namespace seqan3
+{
+
+/*!\brief A gap decorator allows the annotation of sequences with gap symbols
+ *        while leaving the underlying sequence unmodified.
+ * \tparam inner_type The type of range that will be decorated with gaps; must model std::ranges::RandomAccessRange
+ *                    and std::ranges::SizedRange.
+ * \implements seqan3::aligned_sequence_concept
+ * \ingroup decorator
+ *
+ * \details
+ *
+ * This class may be used whenever you want to store or compute an alignment. The underlying (ungapped) sequence
+ * remains unmodified, and is augmented with gap information. The seqan3::gap_decorator_anchor_set behaves just like a
+ * vector over a gapped alphabet when iterating over it, inserting/erasing gaps or accessing a position. The only
+ * difference lies in the performance and size overhead (see below).
+ *
+ *
+ * ### Performance
+ *
+ * **n** The length of the underlying sequence.
+ * **k** The number of contiguous gaps (not gap symbols).
+ * **l** The total number of gap symbols.
+ *
+ * |            | access next | random access    | gap insert/erase at end | gap insert/erase random | size overhead |
+ * |------------|-------------|----------------- |-------------------------|-------------------------|---------------|
+ * | decorator  | \f$O(1)\f$  | \f$O(\log(k))\f$ | \f$O(\log(k))\f$        | \f$O(k)\f$              | O(k)          |
+ * | vector     | \f$O(1)\f$  | \f$O(1)\f$       | \f$O(n)\f$              | \f$O(n)\f$              | O(n)          |
+ *
+ * The *size overhead* refers to the additional space that is needed when using each of the data structures.
+ *
+ * ### Implementation details
+ *
+ * This decorator stores a std::set over tuples of `(pos, cumulative_size)` where every entry represents one
+ * contiguous stretch of gaps. `pos` is the (virtual) insert position in the underlying range and `cumulative_size`
+ * is the length of that contiguous stretch of gaps plus the length of all preceding elements.
+ * Resolving random access requires logarithmic access into the set and inserting or removing a gap symbol additionally
+ * entails updating all subsequent elements in the set to preserve correct cumulative sizes.
+ *
+ */
+template <std::ranges::RandomAccessRange inner_type>
+    requires std::ranges::SizedRange<inner_type>
+class gap_decorator_anchor_set
+{
+private:
+    //!\brief The iterator that moves over the seqan3::gap_decorator_anchor_set.
+    class gap_decorator_anchor_set_iterator
```

Comment:
Please add a details section that contains something like the following sentence:

```
This iterator returns values when dereferenced, not references, i.e. it does not satisfy the semantic requirements of [LegacyForwardIterator](https://en.cppreference.com/w/cpp/named_req/ForwardIterator). It does model the C++20 std::BidirectionalIterator (and std::ForwardIterator implicitly).
```